File: src/main/哈希表/t49/groupAnagrams.go
package main

import (
	"fmt"
	"sort"
)

func groupAnagrams(strs []string) [][]string {
	if strs == nil || len(strs) == 0 {
		return [][]string{}
	}
	n := len(strs)
	getKey := func(byteArray []byte) string {
		sort.Slice(byteArray, func(i, j int) bool {
			return byteArray[i] < byteArray[j]
		})
		return string(byteArray)
	}
	hashMap := make(map[string][]string)
	for i := 0; i < n; i++ {
		bytes := []byte(strs[i])
		key := getKey(bytes)
		hashMap[key] = append(hashMap[key], strs[i])
	}
	res := make([][]string, 0, len(hashMap))
	for _, v := range hashMap {
		res = append(res, v)
	}
	return res
}
func main() {
	strs := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	fmt.Printf("%v", groupAnagrams(strs))

}

File: src/main/哈希表/t49/GroupAnagrams.java
package 哈希表.t49;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * @auther: sotowang
 * @date: 2019/11/23 19:52
 */
public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs == null || strs.length == 0) {
            return new ArrayList<>();
        }
        HashMap<String, List<String>> hashMap = new HashMap<>();
        for (int i = 0; i < strs.length; i++) {
            //将字符排序
            char[] array = strs[i].toCharArray();
            Arrays.sort(array);
            //映射到key
            String key = String.valueOf(array);
            //添加到对应的类中
            if (hashMap.containsKey(key)) {
                hashMap.get(key).add(strs[i]);
            } else {
                List<String> list = new ArrayList<>();
                list.add(strs[i]);
                hashMap.put(key, list);
            }
        }
        return new ArrayList<>(hashMap.values());
    }

    public static void main(String[] args) {
        String[] strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
        System.out.println(new GroupAnagrams().groupAnagrams(strs));

    }
}
